EpsilonAuxBMS/Src/SetContactorsTask.c

```diff
@@ -0,0 +1,242 @@
+#include "SetContactorsTask.h"
+
+static const uint32_t AUX_SET_CONTACTOR_FREQ = 1000; // Every 1 second
+static const float CURRENT_SENSE_RESISTOR  = 0.001; //actually 1 mOhm, but 1 Ohm for testing
+static const int GAIN  = 250;
+static const uint32_t SENSE_SETUP_TIME  = 50; // Setup time for current_sense
+static const float CURRENT_LOWER_THRESHOLD  = 0.15; // Lower current threshold
+static const uint32_t ADC_POLL_TIMEOUT  = 10;
+static const int MAX_ATTEMPTS  = 2; // max attempts for trying to turn on a contactor
+
+typedef enum State {FIRST_CHECK, COMMON_CONTACTOR_ON, CHARGE_CONTACTOR_ON, DISCHARGE_CONTACTOR_ON,
+                    DONE, BLOCKED
+                   } State;
+
+void setContactorsTask(void const* arg)
+{
+    // For concurrency with updateChargeStatusTask()
+    osMutexId* auxStatusMutex = (osMutexId*) arg;
+    // One time osDelayUntil initialization
+    uint32_t prevWakeTime = osKernelSysTick();
+
+    int attemptCount = 0; // Keep track of how many attempts there were to turn on a contactor
+    State state = FIRST_CHECK;
+    State prev_state;
+
+    for (;;)
+    {
+        osDelayUntil(&prevWakeTime, AUX_SET_CONTACTOR_FREQ);
+
+        if (osMutexWait(auxStatusMutex, 0) != osOK)
+        {
+            continue;
+        }
+
+        if (attemptCount > MAX_ATTEMPTS) // If it's greater than a certain number of attempts, report to CAN that something is wrong
```

Comment:
Having the number of attempts intertwined into this function is pretty hard to follow. It'd be better to just have a `setContactors()` which could return an error, and for every error the function gets run again.